### PR TITLE
ConstructorExpressions instead of Macros, No Describe Statement

### PR DESCRIPTION
**Pull request by DerGuteWolf**
Fixes Issue #6 
Provides Compatibility with Steampunk
Remove Compativility with < 740

### CODE REVIEW

File: src/zcl_mustache_parser.clas.testclasses.abap [hunk 1]
```diff
@@ -105,14 +93,25 @@ class ltcl_mustache_parser implementation.
...
+    ).
+
+*    _add_mu_token lt_exp zif_mustache=>c_token_type-static      ''  1   `Our sales:`.
+*    _add_mu_token lt_exp zif_mustache=>c_token_type-static      ''  1   c_nl.
+*    _add_mu_token lt_exp zif_mustache=>c_token_type-section     '=' 1   'items'.
+*    _add_mu_token lt_exp zif_mustache=>c_token_type-static      ''  2   `* `.
+*    _add_mu_token lt_exp zif_mustache=>c_token_type-etag        ''  2   'name'.
+*    _add_mu_token lt_exp zif_mustache=>c_token_type-static      ''  2   ` - $`.
+*    _add_mu_token lt_exp zif_mustache=>c_token_type-etag        ''  2   'price'.
+*    _add_mu_token lt_exp zif_mustache=>c_token_type-static      ''  2   c_nl.
```
Comment (sbcgua) [inline src/zcl_mustache_parser.clas.testclasses.abap]: Could you remove the comments pls ?

File: src/zcl_mustache_render.clas.abap [hunk 2]
```diff
@@ -128,8 +128,9 @@ CLASS ZCL_MUSTACHE_RENDER IMPLEMENTATION.
                       it_data_stack = it_data_stack
                       iv_level      = lines( it_data_stack ) ). " Start from deepest level
 
+    DATA(t) = cl_abap_datadescr=>describe_by_data_ref( lr ).
     assign lr->* to <field>.
-    describe field <field> type lv_type.
+    lv_type = cl_abap_datadescr=>get_data_type_kind( <field> ).
```
Comment (sbcgua) [inline src/zcl_mustache_render.clas.abap:133]: Hmmm, the describe is in fact faster if I remember correctly, that's why it was choosen in the first place.

**Review by sbcgua — CHANGES_REQUESTED** [review]
See some comments below.

Is there a need to make it above 740 ? I'm not using it in older systems but someone can potentially ... so I'm a bit tentative ...

**Comment by larshp** [comment]
@sbcgua will you be dropping 702 compatibility?

**Comment by sbcgua** [comment]
@larshp I'm prone **not** to. But I'm not using this lib currently in my projects in our old system. So ... it depends on who uses it and where. How to find it out ? Maybe we can move 740 support into a branch ?

@DerGuteWolf Could you pls tell what is your reasoning behind this PR ?

**Comment by larshp** [comment]
it might be a good example for automatically downporting using abaplint if needed

**Comment by sbcgua** [comment]
> it might be a good example for automatically downporting using abaplint if needed

@larshp 
I think it is a good case. Let's make a branch for 740 then after this PR is merged. The above changes are not automatable by the way - it is a replacement of macros.

**Comment by DerGuteWolf** [comment]
describe is not allowed in steampunk, therefore the reason to remove this. Also Macros are not allowed in steampunk, therefore I substituted macros with VALUE construction expressions

File: src/zcl_mustache_render.clas.abap (hunk 2 above)
Comment (DerGuteWolf) [reply src/zcl_mustache_render.clas.abap:133]: describe is not allowed in steampunk

File: src/zcl_mustache_parser.clas.testclasses.abap (hunk 1 above)
Comment (DerGuteWolf) [reply src/zcl_mustache_parser.clas.testclasses.abap]: Comment is removed, sorry, was a leftover

**Comment by sbcgua** [comment]
Thanks !

> describe is not allowed in steampunk, therefore the reason to remove this

Would you mind if I ask you to measure the perf before and after and publish here ? If I remember well the test prog is already created - perf_test.prog - though it might need this - https://github.com/sbcgua/benchmarks - I don't remember for sure

**Comment by DerGuteWolf** [comment]
> Would you mind if I ask you to measure the perf before and after and publish here ? 

Will do this, however I need some shuffling around first.

**Comment by DerGuteWolf** [comment]
> Would you mind if I ask you to measure the perf before and after and publish here ? 

Will do this, however I need some shuffling around first.

**Comment by sbcgua** [comment]
With describe
image

With RTTI
image

40% slower :( eh ...

**Review by sbcgua — APPROVED** [review]

**Comment by sbcgua** [comment]
OK. 
1) Merged into a separate branch - steampunk
2) Added a reference to the main readme - https://github.com/sbcgua/abap_mustache/tree/master#versions

@DerGuteWolf Thanks for the PR !
@larshp feel free to PR experiments with auto up-porting. I guess current UTs are able to withstand that. :)

**Pull request merged**